homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -114,101 +126,104 @@ def next_departuredate(departure):
 
+    _attr_icon = ICON
     _attr_device_class = SensorDeviceClass.TIMESTAMP
 
-    def __init__(self, train_api, name, from_station, to_station, weekday, time):
+    def __init__(
+        self,
+        train_api: TrafikverketTrain,
+        name: str,
+        from_station: str,
+        to_station: str,
+        weekday: list,
+        departuretime: time,
+    ) -> None:
         """Initialize the sensor."""
         self._train_api = train_api
-        self._name = name
+        self._attr_name = name
         self._from_station = from_station
         self._to_station = to_station
         self._weekday = weekday
-        self._time = time
-        self._state = None
-        self._departure_state = None
-        self._delay_in_minutes = None
-        self._timezone = get_time_zone("Europe/Stockholm")
+        self._time = departuretime
 
-    async def async_update(self):
+    async def async_update(self) -> None:
         """Retrieve latest state."""
-        if self._time is not None:
+        when = datetime.now()
+        _state: TrainStop | None = None
+        if self._time:
             departure_day = next_departuredate(self._weekday)
-            when = datetime.combine(departure_day, self._time).astimezone(
-                self._timezone
-            )
-            try:
-                self._state = await self._train_api.async_get_train_stop(
+            when = datetime.combine(departure_day, self._time).astimezone(TIMEZONE)
+        try:
+            if self._time:
+                _state = await self._train_api.async_get_train_stop(
                     self._from_station, self._to_station, when
                 )
-            except ValueError as output_error:
-                _LOGGER.error(
-                    "Departure %s encountered a problem: %s", when, output_error
+            else:
+
+                _state = await self._train_api.async_get_next_train_stop(
+                    self._from_station, self._to_station, when
                 )
-        else:
-            when = datetime.now()
-            self._state = await self._train_api.async_get_next_train_stop(
-                self._from_station, self._to_station, when
+        except ValueError as output_error:
+            _LOGGER.error("Departure %s encountered a problem: %s", when, output_error)
+
+        if not _state:
+            self._attr_available = False
+            self._attr_native_value = None
+            self._attr_extra_state_attributes = {}
+            return
+
+        self._attr_available = True
+
+        self._attr_native_value = _state.advertised_time_at_location.astimezone(
+            TIMEZONE
+        )
+        if _state.time_at_location:
+            self._attr_native_value = _state.time_at_location.astimezone(TIMEZONE)
+        if _state.estimated_time_at_location:
+            self._attr_native_value = _state.estimated_time_at_location.astimezone(
+                TIMEZONE
             )
```

Comment:
For the native state values, we don't need to convert timezones.
If the original datetime object has a timezone already, we can just use the original.

Reply:
The original datetime doesn't provide a timezone so therefore attaching it here.

Reply:
Can you maybe add a comment to that effect in the code?
Maybe rename the variable as STOCKHOLM_TIMEZONE also.

Reply:
Have a look if the comment seems fine, also adjusted the name of the timezone as suggested

Reply:
I'm leaving as unresolved for @frenck

Reply:
If you want to attach a timezone, you should use `.replace(tzinfo=TIMEZONE)`, not `astimezone(TIMEZONE)`.

Reply:
Done 👍